File: downloads/code/LeetCode/Java/Binary-Tree-Right-Side-View.java
/*
 *  Definition for a binary tree node.
 *  public class TreeNode {
 *      int val;
 *      TreeNode left;
 *      TreeNode right;
 *      TreeNode(int x) { val = x; }
 *      }
 */
public class Solution {
        public List<Integer> rightSideView(TreeNode root) {
            List<Integer> ret = new LinkedList<Integer>();
            if(root==null)
                return ret;
            helper(ret, root, 0);
            return ret;
        }

        public void helper(List<Integer> ret, TreeNode node, int depth){
            if(node == null)
                return ;
            if(depth == ret.size()){
                ret.add(node.val);
            }

            helper(ret, node.right, depth+1);
            helper(ret, node.left, depth+1);
        
        }
}

File: downloads/code/LeetCode/Python/Binary-Tree-Right-Side-View.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """

        stack = []
        ret = []

        if root == None:
            return ret

        stack.append([root,level])

        while stack:
            top = stack.pop(0)
            node = top[0]
            cur_level = top[1]
            
            if node.left != None:
                stack.append([node.left,cur_level+1])
            if node.right != None:
                stack.append([node.right,cur_level+1])
            
            if not stack or cur_level != stack[0][1]:
                ret.append(node.val)

        return ret
